rust/rust.bzl

```diff
@@ -529,65 +277,65 @@ def _rust_doc_impl(ctx):
     doc_flags = _build_rustdoc_flags(ctx)
 
     # Build rustdoc command.
-    doc_cmd = build_rustdoc_command(ctx, toolchain, rust_doc_zip, depinfo, lib_rs, target, doc_flags)
+    doc_cmd = build_rustdoc_command(toolchain, rust_doc_zip, depinfo, crate, doc_flags)
 
     # Rustdoc action
-    rustdoc_inputs = (target.srcs +
-                      depinfo.transitive_libs + [toolchain.rust_doc] + toolchain.rustc_lib +
-                      toolchain.rust_lib)
+    rustdoc_inputs = (
+        crate.srcs +
+        depinfo.transitive_libs +
+        [toolchain.rust_doc] +
+        toolchain.rustc_lib +
+        toolchain.rust_lib
+    )
 
     ctx.action(
         inputs = rustdoc_inputs,
         outputs = [rust_doc_zip],
         mnemonic = "Rustdoc",
         command = doc_cmd,
         use_default_shell_env = True,
-        progress_message = "Generating rustdoc for {} ({} files)".format(
-                            target.name, len(target.srcs)),
+        progress_message = "Generating rustdoc for {} ({} files)".format(crate.name, len(crate.srcs)),
     )
 
 def _rust_doc_test_impl(ctx):
     """Implementation for the rust_doc_test rule."""
-    rust_doc_test = ctx.outputs.executable
 
-    # Gather attributes about the rust_library target to generated rustdocs for.
-    target = struct(
-        name = ctx.label.name,
-        srcs = ctx.attr.dep.rust_srcs,
-        crate_root = ctx.attr.dep.crate_root,
-        deps = ctx.attr.dep.rust_deps,
-    )
+    if not hasattr(ctx.attr.dep, "crate_info"):
+        fail("Expected rust library or binary.", "dep")
 
-    # Find lib.rs
-    lib_rs = (_find_crate_root_src(target.srcs, ["lib.rs", "main.rs"]) if target.crate_root == None else target.crate_root)
+    crate = ctx.attr.dep.crate_info
+    rust_doc_test = ctx.outputs.executable
 
     toolchain = _find_toolchain(ctx)
 
-    # Get information about dependencies
-    output_dir = rust_doc_test.dirname
     working_dir = "."
     depinfo = _setup_deps(
-        target.deps,
-        target.name,
+        [ctx.attr.dep],
```

Comment:
This bit was pretty off, setup was wrong and none of it was being exercised because rustdoc wasn't being called with --test